relay-server/src/actors/outcome.rs

```diff
@@ -622,31 +704,144 @@ impl Actor for HttpOutcomeProducer {
     type Context = Context<Self>;
 }
 
-impl Supervised for HttpOutcomeProducer {}
+impl Handler<TrackRawOutcome> for HttpOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
 
-impl SystemService for HttpOutcomeProducer {}
+    fn handle(&mut self, msg: TrackRawOutcome, ctx: &mut Self::Context) -> Self::Result {
+        self.send_http_message(msg, ctx);
+        Ok(())
+    }
+}
 
-impl Default for HttpOutcomeProducer {
+impl Handler<TrackOutcome> for HttpOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
+
+    fn handle(&mut self, msg: TrackOutcome, ctx: &mut Self::Context) -> Self::Result {
+        self.handle(TrackRawOutcome::from_outcome(msg, &self.config), ctx)
+    }
+}
+
+struct ClientReportOutcomeProducer {}
+
+impl Actor for ClientReportOutcomeProducer {
+    type Context = Context<Self>;
+}
+
+impl Handler<TrackOutcome> for ClientReportOutcomeProducer {
+    type Result = Result<(), OutcomeError>;
+
+    fn handle(&mut self, msg: TrackOutcome, _ctx: &mut Self::Context) -> Self::Result {
+        let mut client_report = ClientReport {
+            timestamp: Some(UnixTimestamp::from_secs(
+                msg.timestamp.timestamp().try_into().unwrap_or(0),
+            )),
+            ..Default::default()
+        };
+
+        // The outcome type determines what field to place the outcome in:
+        let discarded_events = match msg.outcome {
+            Outcome::Filtered(_) => &mut client_report._server_filtered_events,
+            Outcome::FilteredSampling(_) => &mut client_report._server_filtered_sampling_events,
+            Outcome::RateLimited(_) => &mut client_report._server_rate_limited_events,
+            _ => {
+                // Cannot convert this outcome to a client report.
+                return Ok(());
+            }
+        };
+
+        // Now that we know where to put it, let's create a DiscardedEvent
+        let discarded_event = DiscardedEvent {
+            reason: msg.outcome.to_reason().unwrap_or_default().to_string(),
+            category: msg.category,
+            quantity: msg.quantity,
+        };
+        discarded_events.push(discarded_event);
+
+        relay_log::trace!("Sending client report");
+        // Assuming it is not necessary to batch requests here (like HttpOutcomeProducer does),
+        // because outcomes were already aggregated by `OutcomeAggregator`
```

Comment:
@untitaker we discussed that we should batch requests here anyway. But it only makes sense to batch per-project, since client reports are sent in envelopes. Do we want to maintain a flush state for each project?

Reply:
I guess we can leave it out for now but my fear is that customer relays will send us too many messages.